WIP: Use _Brain in plot_vector_source_estimates

GuillaumeFavelier opened the pull request:
This PR is a follow-up of #6731 suggested also in [this comment](https://github.com/mne-tools/mne-python/pull/6731#issuecomment-536042199). The goal is to make `stc_vec.plot()` use the `_Brain` class which relies on the abstract 3d backend layer `_Renderer`.

**ToDo**

- [x] Fix scaling of the glyphs by using scalar values (merged [upstream](https://github.com/pyvista/pyvista/pull/451))
- [x] Enable Depth Peeling (merged [upstream](https://github.com/pyvista/pyvista/pull/450))
- [x] Increase arrow thickness
- [x] Fix colorbar range
- [x] Update the 3D tests with the `renderer` fixture (c.f. https://github.com/mne-tools/mne-python/pull/7028#issue-337436335)
- [x] Experiment 3D arrows glyphs (feedback from #7030)
- [x] Colormap has a 'grey tone'
- [x] Refactor smoothing matrix generation
- [x] The left and right hemi data have different scales
- [x] Rendering dark filter bug on CircleCI
- [x] remove glyph options (should be future PR)
- [x] `time_viewer=True` only playback only updates mesh, not vectors
- [x] hitting spacebar at the end of playback aborts Python (should restart from beginning), fix was in the reverted PR
- [x] fix glyph flashing during playback (if still an issue)
- [x] get CIs green
- [x] add support for `show_traces=True` if it's not too difficult
- [x] fix smoothing steps bug

It's an item of #7162

GuillaumeFavelier commented:
This is my local result using the [gist](https://gist.github.com/rythorpe/c51a5b66b492b67437843283c67050af) suggested in [this comment](https://github.com/mne-tools/mne-python/issues/7020#issue-517400092). This is still work in progress.

![2019-11-14_1920x1080](https://user-images.githubusercontent.com/18143289/68877189-f2ebec00-0705-11ea-8421-37327c4fb55c.png)

GuillaumeFavelier commented:
Better to wait for the merge of #7084, I expect some conflicts to solve.

GuillaumeFavelier commented:
Meanwhile, these are my results: 

Mayavi | PyVista
---------|----------
![2019-11-20_1920x1080](https://user-images.githubusercontent.com/18143289/69249977-1b189680-0baf-11ea-8e60-f80372d38679.png) | ![2019-11-20_1920x1080_000](https://user-images.githubusercontent.com/18143289/69250122-6468e600-0baf-11ea-91c8-af414646640f.png)

Note: this "grey tone" is still here. I suspect there is some scaling of the colormap to do.

larsoner commented:
Apparently no conflicts after merging #7084, but feel free to rebase anyway @GuillaumeFavelier

GuillaumeFavelier commented:
I think we don't see the conflicts because Azure is failing. I see them locally though

larsoner commented:
@GuillaumeFavelier any chance to get to this in the next two weeks to get this into 0.20? That way we can have a complete replacement for 3D viz (outside of `mne coreg`), right?

GuillaumeFavelier commented:
I'll increase the priority of this PR but after so long the same blockers remain:

* The glyph size is not consistent with Mayavi
* Some kind of grey filter is applied on the colorbar
* The renders on sphinx-gallery are far from satisfying

larsoner commented:
Okay I'll move the milestone to 0.21 then

GuillaumeFavelier commented:
Eventually I think this could be merged but the overview table in `mne/viz/backends/renderer.py` should be updated with only partial support for `plot_vector_source_estimates` with an additional note saying what's missing or not working.

GuillaumeFavelier commented:
I notice a very big lag during playback when `time_viewer=True` and I think the memory consumption is huge especially when `hemi='both'`. Did I do something wrong with the interpolation?

larsoner commented:
Is it spatially upsampling the entire n_vertices x n_times array (bad), or upsampling a single time point at a time (good) ?

GuillaumeFavelier commented:
I suppose it's bad so I'll try "upsampling a single time point at a time"

larsoner commented:
> I suppose it's bad so I'll try "upsampling a single time point at a time"

It looked like the only place the upsampling of the full was being done was to get the max -- and this you can compute on the data itself. Pushed a commit to do this. Feel free to try it and see if things are still laggy. Hopefully they aren't!

larsoner commented:
Also pushed a commit to touch probably relevant examples

larsoner commented:
I'm going to re-mark this for 0.20 and we can scrap it if need be

larsoner commented:
Looks like it's getting close @GuillaumeFavelier! Updated examples:

- tutorials/source-modeling/plot_dipole_orientations.py [PR](https://18783-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_dipole_orientations.html#loose-dipole-orientations) / [master](https://mne.tools/dev/auto_tutorials/source-modeling/plot_dipole_orientations.html#loose-dipole-orientations)
- tutorials/source-modeling/plot_visualize_stc.py [PR](https://18783-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_visualize_stc.html#vector-source-estimates) / [master](https://mne.tools/dev/auto_tutorials/source-modeling/plot_visualize_stc.html#vector-source-estimates)
- examples/inverse/plot_vector_mne_solution.py [PR](https://18783-1301584-gh.circle-artifacts.com/0/dev/auto_examples/inverse/plot_vector_mne_solution.html#sphx-glr-auto-examples-inverse-plot-vector-mne-solution-py) / [master](https://mne.tools/dev/auto_examples/inverse/plot_vector_mne_solution.html#sphx-glr-auto-examples-inverse-plot-vector-mne-solution-py)

By eye all three of these are getting close but suffer from the same problems:

- some rendering problems, maybe blending modes or something, probably worth checking what PySurfer does?
- colorbar is wrong

GuillaumeFavelier commented:
> some rendering problems, maybe blending modes or something, probably worth checking what PySurfer does?

I investigated for some time now and I can't seem to reproduce *exactly* what PySurfer does. I noticed the rendering issues too but I have no solution for that.

> colorbar is wrong

I'll see what I can do

GuillaumeFavelier commented:
Strangely enough, the grey tone on the colorbar is fixed when `time_viewer=True`...

GuillaumeFavelier commented:
The grey tone and the range of the colorbar are fixed now.

GuillaumeFavelier commented:
Regression detected in: 

https://18790-1301584-gh.circle-artifacts.com/0/dev/auto_tutorials/source-modeling/plot_dipole_orientations.html#sphx-glr-auto-tutorials-source-modeling-plot-dipole-orientations-py

I have to fix this

GuillaumeFavelier commented:
The order of certain drawing operations are important to achieve the correct render, I commented to make it clear.

larsoner commented:
Style is unhappy

Does it render properly now? If so, what else is left to do?

larsoner commented on mne/viz/_brain/_brain.py:
```diff
@@ -453,6 +450,36 @@ def time_label(x):
+            self._data[hemi]['actor'].append(actor)
+            self._data[hemi]['mesh'].append(mesh)
+
+        # 2) udpate time and smoothing properties
```
udpate -> update

GuillaumeFavelier commented:
The colorbar is fixed, this:

> some rendering problems, maybe blending modes or something

is not fixed. I updated the note about it in `mne/viz/backends/renderer`

GuillaumeFavelier commented:
Waiting on the CIs to confirm

larsoner commented:
> is not fixed. I updated the note about it in mne/viz/backends/renderer

Can you update the top comment? It seems like that's out of date now. It would be good to know what still needs to be done. I might have time to look into the rendering bugs today

GuillaumeFavelier commented:
It's updated now @larsoner

larsoner commented:
After having read through the code that PySurfer uses, I'm convinced it's wrong because it scales the left and right hemisphere data potentially differently, unless it's compensating for something that Mayavi does under the hood. This will require more investigation, but if it's the case we should probably fix the behavior somehow :(

larsoner commented:
In the meantime I pushed a commit that unifies how/where smoothing matrices are calculated by refactoring slightly and reusing code from `mne/morph.py` rather than `mne/viz/_brain/utils.py`

GuillaumeFavelier commented on mne/viz/_brain/_brain.py:
```diff
@@ -486,15 +487,6 @@ def time_label(x):
 
-                if scale_factor is None:
-                    width = np.ptp(self.geo[hemi].coords[:, 1])
-                    final_scale_factor = width * 0.1
```
We could iterate through the hemis here to find a suitable scale factor

GuillaumeFavelier commented:
With the latest version of `quiver3d()`, the following script gives the same result for both `mayavi` and `pyvista`:

```py
import mne

for backend in ('mayavi', 'pyvista'):
    mne.viz.set_3d_backend(backend)
    from mne.viz.backends.renderer import _get_renderer
    # scale by vector values
    renderer = _get_renderer()
    renderer.quiver3d([0, 0, 0], [0, 0, 0], [0, 0, 0],
                      [0, 0, 5], [10, 0, 0], [0, 25, 0],
                      color="red", scale=2., scale_mode='vector',
                      mode="2darrow")
    renderer.show()

    # scale by scalar values
    renderer = _get_renderer()
    renderer.quiver3d([0, 0, 0], [0, 0, 0], [0, 0, 0],
                      [0, 0, 1], [1, 0, 0], [0, 1, 0],
                      scalars=[10, 25, 5],
                      color="red", scale=2., scale_mode='scalar',
                      mode="2darrow")
    renderer.show()
```

agramfort commented:
ready to test?

>

GuillaumeFavelier commented:
I'm still not satisfied by this version @agramfort. I tried to work directly on `quiver3d` to obtain something backend-agnostic but either in the parameters sent by `PySurfer` or the post-processing on the glyphs, there is something that remains mysterious to me that I can't quite reproduce with `PyVista`.

GuillaumeFavelier commented:
I believe 8073aef should reproduce https://github.com/nipy/PySurfer/blob/master/surfer/viz.py#L2049-L2061 and fix the color issues on the glyphs.

larsoner commented:
@GuillaumeFavelier let me know when you're tired of trying and I'll take a look

GuillaumeFavelier commented:
@larsoner I pushed a commit to fix the warnings in the CIs, feel free to take a look

GuillaumeFavelier commented:
@larsoner I pushed a commit to enable depth peeling on pyvista. It's especially useful for the users of `vtk<8.2` because after this, it's enabled by default

larsoner commented on mne/viz/_3d.py:
```diff
@@ -2365,6 +2365,10 @@ def plot_vector_source_estimates(stc, subject=None, hemi='lh', colormap='hot',
+    glyph : '2darrow' | '3darrow'
+        The type of glyphs to use. If '2darrow', the glyphs belong to a plan
+        and if '3darrow', the glyphs are composite 3d objects (cone +
+        cylinder).
```
@GuillaumeFavelier can we just get rid of this option for now? I'm not sure it's going to work given all the scaling issues

GuillaumeFavelier replied:
It should not be difficult to add later. I'll remove it for now because it's not top priority

GuillaumeFavelier commented:
The flickering of the glyphs happens because the glyphs actor is removed and re-added in the scene. So if I update the geometry of the glyphs already present in the scene without adding a new actor, it should be fixed.

larsoner commented:
Are you working on that now, or is it implemented already?

GuillaumeFavelier commented:
I'm pushing right now

GuillaumeFavelier commented:
This is what I obtain:

![output](https://user-images.githubusercontent.com/18143289/77425197-a2328100-6dd2-11ea-9290-f212e3fdfecd.gif)

larsoner commented:
Awesome!

Want to merge with master and try show_traces='auto'?

GuillaumeFavelier commented:
I think this PR fails on azure somehow

larsoner commented:
Looks like the failure occurs in `test_report.py`, or at least that's the test that should be next in the order. Could something here have broken it?

agramfort commented:
this works great now for me !

can we fix conflict and merge when CIs are green?

GuillaumeFavelier commented:
I'll investigate the failures

GuillaumeFavelier commented:
Waiting on Travis here then I'll work on Azure

larsoner commented on mne/source_estimate.py:
```diff
@@ -2246,7 +2246,7 @@ def __init__(self, data, vertices=None, tmin=None, tstep=None,
                      colormap='auto', time_label='time=%02.f ms',
                      smoothing_steps=10,
-                     transparent=None, alpha=1.0, time_viewer=False,
+                     transparent=None, alpha=1.0, time_viewer='auto',
```
These will also need to be changed in `_3d.py`

GuillaumeFavelier replied:
This was done in #7380 no?

larsoner replied:
Ahh, yes I guess it was

GuillaumeFavelier commented:
The failures don't seem related to depth peeling, I'll have to split this PR to have a better perspective

larsoner commented:
Killing CIs for now while we wait for #7513

GuillaumeFavelier commented:
Okay

larsoner commented:
... actually I'll just bring the optimistic fix here

GuillaumeFavelier commented:
Checking Azure here then

GuillaumeFavelier commented:
It seems it's indeed related to AA. And you think it would not work on Azure if we remove:

```py
    if os.getenv('AZURE_CI_WINDOWS', 'false').lower() == 'true':
        return
```

Now that we check if the figure is active?

larsoner commented on mne/viz/_3d.py:
```diff
@@ -2391,9 +2400,18 @@ def plot_vector_source_estimates(stc, subject=None, hemi='lh', colormap='hot',
+            not using_mayavi and
+            time_viewer and
+            # XXX temporary hidden workaround for memory problems on CircleCI
+            os.getenv('_MNE_BRAIN_TRACES_AUTO', 'true').lower() != 'false')
```
This looks like code dup that should be refactored with `plot_source_estimates`

larsoner commented:
> It seems it's indeed related to AA. And you think it would not work on Azure if we remove:

The old check was just to see if `ren_win` existed. All the PyVista code does is set some `ren_win` things and call `modified`. So it should have errored out if `ren_win` did not exist.

So I think `is_active` might not be checking that things are sufficiently activated (e.g., OpenGL context actually created) to set the necessary properties.

larsoner commented:
... oh I guess it does not actually check `ren_win` (the depth peeling code does). So maybe it would be sufficient. I guess you could try it, but it might be better to put effort into getting this PR merged with green CIs, then undo some of the workarounds if possible in later PRs.

GuillaumeFavelier commented:
> getting this PR merged with green CIs, then undo some of the workarounds if possible in later PRs.

+100, this PR is up for too long

GuillaumeFavelier commented:
> This looks like code dup that should be refactored with plot_source_estimates

Sorry for being paranoid but I'll push my latest changes once the CIs come back green.

larsoner commented:
Awesome that `time_viewer='auto'` is now working!

@GuillaumeFavelier another fix that I think was in the reverted PR that I just noticed was that `smoothing_steps` now gets overridden by the slider value instead of the slider being set to whatever value was passed in `smoothing_steps`. You can replicate by passing for example `smoothing_steps=1` which very clearly have gaps, it shows up then gets "snapped" up to 7 steps.

Also this suggests that the steps are being computed twice during init, which will slow things down.

If you fix this, please also check that `smoothing_steps='nearest'` works properly, because we have to be careful about converting that to `0` when setting the slider value.

Then I think we're ready for merge, right?

GuillaumeFavelier commented:
I fixed the bugs with `smoothing_steps` and updated the tests. I also removed the duplicated calls to `set_data_smoothing` and `set_time_point` in `_TimeViewer`.

Looking at the todo list and assuming the CIs come back green, I think we're ready now.

agramfort commented:
still works great for me !

>

GuillaumeFavelier commented:
> still works great for me !

Very reassuring @agramfort 😅

larsoner commented:
Works great! Thanks for the awesome work @GuillaumeFavelier
